File: library/src/main/java/com/handy/basic/mvp/BasePresenter.java
package com.handy.basic.mvp;

import android.arch.lifecycle.Lifecycle;
import android.arch.lifecycle.LifecycleOwner;
import android.arch.lifecycle.OnLifecycleEvent;
import android.support.annotation.NonNull;

/**
 * Mvp框架中 Presenter层通用基类
 *
 * @author LiuJie https://github.com/Handy045
 * @description 所有Presenter层均需要实现此接口
 * @date Created in 2018/2/1 上午11:16
 * @modified By LiuJie
 */
public class BasePresenter implements IPresenter {

    private LifecycleOwner lifecycleOwner;

    public BasePresenter(@NonNull LifecycleOwner lifecycleOwner) {
        this.lifecycleOwner = lifecycleOwner;
        onStart();
    }

    /**
     * 启动
     */
    @Override
    public void onStart() {
        //将 LifecycleObserver 注册给 LifecycleOwner 后 @OnLifecycleEvent 才可以正常使用
        if (lifecycleOwner != null) {
            lifecycleOwner.getLifecycle().addObserver(this);
        }
    }

    /**
     * 销毁
     */
    @Override
    public void onDestroy() {
        this.lifecycleOwner = null;
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    void onDestroy(LifecycleOwner owner) {
        owner.getLifecycle().removeObserver(this);
    }
}

File: library/src/main/java/com/handy/basic/mvp/BasePresenter.kt
package com.handy.basic.mvp

import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.OnLifecycleEvent

/**
 * @title: BasePresenter
 * @projectName HandyBasicKT
 * @description: Mvp-Presenter通用超类
 * @author LiuJie https://github.com/Handy045
 * @date Created in 2019-10-11 16:30
 */
open class BasePresenter(private var lifecycleOwner: LifecycleOwner?) : IPresenter {
    init {
        // 将 LifecycleObserver 注册给 LifecycleOwner 后 @OnLifecycleEvent 才可以正常使用
        lifecycleOwner?.lifecycle?.addObserver(this)
    }

    override fun onDestroy() {
        this.lifecycleOwner = null
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    internal fun onDestroy(owner: LifecycleOwner) {
        owner.lifecycle.removeObserver(this)
    }
}